### PR TITLE
Spark TaskContext

### PR DESCRIPTION
**Pull request by raver119**
As we've recently found, we have issue with background threads using Spark Iterator<DataSet> etc.

Problem is Spark read/write lock mechanics. RDD blocks are considered locked by accessor's TaskId, instead of Java ThreadId. And if some thread accessing RDD has no TaskContext set - common TaskId of -1024 is used. And this makes race conditions possible, throwing exceptions like this:

```
Exception in thread "ADSI prefetch thread" java.lang.AssertionError: assertion failed: Task -1024 release lock on block rdd_0_0 more times than it acquired it
	at scala.Predef$.assert(Predef.scala:165)
	at org.apache.spark.storage.BlockInfoManager.unlock(BlockInfoManager.scala:298)
	at org.apache.spark.storage.BlockManager.releaseLock(BlockManager.scala:654)
	at org.apache.spark.storage.BlockManager$$anonfun$1.apply$mcV$sp(BlockManager.scala:458)
	at org.apache.spark.util.CompletionIterator$$anon$1.completion(CompletionIterator.scala:46)
	at org.apache.spark.util.CompletionIterator.hasNext(CompletionIterator.scala:35)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:39)
	at scala.collection.convert.Wrappers$IteratorWrapper.hasNext(Wrappers.scala:30)
	at org.deeplearning4j.datasets.iterator.IteratorDataSetIterator.next(IteratorDataSetIterator.java:54)
	at org.deeplearning4j.datasets.iterator.IteratorDataSetIterator.next(IteratorDataSetIterator.java:44)
	at org.deeplearning4j.datasets.iterator.IteratorDataSetIterator.next(IteratorDataSetIterator.java:18)
	at org.deeplearning4j.datasets.iterator.AsyncDataSetIterator$AsyncPrefetchThread.run(AsyncDataSetIterator.java:407)
```

So, this PR is proposed as hot-fix for this issue. When prefetch iterators are created, we get TaskContext from current thread and propagate it to new background thread. Since we're strictly following thread-safety of DataSetIterator use, we won't break anything in Spark locks mechanics.

/cc @agibsonccc @AlexDBlack @huitseeker 

P.s. proposed PR compiles/works fine for both 1.6 and 2.1

### CODE REVIEW

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/scala/org/apache/spark/TaskContextHelper.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package org.apache.spark
+
+
+/**
+  * This simple helper is used to get access to package-protected Scala TaskContext.setTaskContext method
```
Comment (huitseeker) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/scala/org/apache/spark/TaskContextHelper.scala]: I would like to see an explicit reference to https://issues.apache.org/jira/browse/SPARK-18406 in comments here.

**Comment by raver119** [comment]
Thanks guys

**Comment by jasstionzyf** [comment]
@AlexDBlack unluck,  after i pull latest snapshot release, i still get the folowing error:
17/07/04 18:47:09 ERROR util.SparkUncaughtExceptionHandler: [Container in shutdown] Uncaught exception in thread Thread[ADSI prefetch thread,5,main]
java.lang.AssertionError: assertion failed: Task -1024 release lock on block rdd_1_29 more times than it acquired it
        at scala.Predef$.assert(Predef.scala:170)
        at org.apache.spark.storage.BlockInfoManager.unlock(BlockInfoManager.scala:298)
        at org.apache.spark.storage.BlockManager.releaseLock(BlockManager.scala:654)
        at org.apache.spark.storage.BlockManager$$anonfun$1.apply$mcV$sp(BlockManager.scala:458)
        at org.apache.spark.util.CompletionIterator$$anon$1.completion(CompletionIterator.scala:46)
        at org.apache.spark.util.CompletionIterator.hasNext(CompletionIterator.scala:35)
        at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:39)
        at scala.collection.convert.Wrappers$IteratorWrapper.hasNext(Wrappers.scala:30)
        at org.deeplearning4j.datasets.iterator.IteratorDataSetIterator.next(IteratorDataSetIterator.java:54)
        at org.deeplearning4j.datasets.iterator.IteratorDataSetIterator.next(IteratorDataSetIterator.java:44)

**Comment by jasstionzyf** [comment]
i find a problem ,  for this push,  all pom file changes have not push into master branch

**Comment by raver119** [comment]
Sorry, what do you mean?

**Comment by jasstionzyf** [comment]
@raver119 i means when i clone the master branch from git,  and run git pull,  all the pom has not changed as this request , but java class have changed

**Comment by raver119** [comment]
But that's expected behavior, we have special scripts for that purpose in the root folder of the repo. Well, multiple scripts: change Scala version, and change Spark version.

For nd4j there's also similar script that changes target CUDA version.

**Pull request merged**